hr_timesheet_rounded/models/account_analytic_line.py

```diff
@@ -0,0 +1,138 @@
+                rec.project_id.timesheet_rounding_granularity,
+                rec.project_id.timesheet_rounding_method,
+                rec.project_id.timesheet_invoicing_factor,
+                rec.unit_amount,
+            )
+
+    @staticmethod
+    def _calc_rounded_amount(granularity, rounding_method, factor, amount):
+        factor = factor / 100.0
+        if granularity:
+            unit_amount_rounded = float_round(
+                amount * factor,
+                precision_rounding=granularity,
+                rounding_method=rounding_method
+            )
+        else:
+            unit_amount_rounded = amount * factor
+        return unit_amount_rounded
+
+    def _timesheet_postprocess(self, values):
+        # TODO: replace this w/ an override of `_timesheet_postprocess_values`
+        # where we update the values passed then to write.
+        # This way we could get rid of `_update_unit_amount_rounded`
+        # and the consequent call to write,
+        # event though it might be still useful for the onchange.
+        # NOTE: we don't have to replace `amount` value.
+        if not self.env.context.get('update_unit_amount_rounded'):
+            if not values.get('unit_amount_rounded'):
+                self._update_unit_amount_rounded()
+        return super()._timesheet_postprocess(values)
+
+    ####################################################
+    # ORM Overrides
+    ####################################################
+    @api.model
+    def read_group(self, domain, fields, groupby, offset=0,
+                   limit=None, orderby=False, lazy=False):
+        """Replace the value of unit_amount by unit_amount_rounded.
+
+        When context key `timesheet_rounding` is True
+        we change the value of unit_amount with the rounded one.
+        This affects `sale_order_line._compute_delivered_quantity`
+        which in turns compute the delivered qty on SO line.
+        """
+        ctx_ts_rounded = self.env.context.get('timesheet_rounding')
+        if ctx_ts_rounded:
+            # To add the unit_amount_rounded value on read_group
+            fields.append('unit_amount_rounded')
+        res = super().read_group(
+            domain, fields, groupby, offset=offset,
+            limit=limit, orderby=orderby, lazy=lazy
+        )
+        if ctx_ts_rounded:
+            # To set the unit_amount_rounded value insted of unit_amount
+            for rec in res:
+                rec['unit_amount'] = rec['unit_amount_rounded']
+        return res
+
+    def read(self, fields=None, load='_classic_read'):
+        """Replace the value of unit_amount by unit_amount_rounded.
+
+        When context key `timesheet_rounding` is True
+        we change the value of unit_amount with the rounded one.
+        This affects `account_anaytic_line._sale_determine_order_line`.
+        """
+        ctx_ts_rounded = self.env.context.get('timesheet_rounding')
+        fields_local = fields[:] if fields else []
```

Comment:
`fields_local = fields or []` ?

Reply:
nope, we want a copy here: if fields is value you get a copy of it else is an empty list

Reply:
Thanks for the explanation.
However, adding it as comment in the code would be better, or using `fields.copy()` would be more explicit. :wink: 
Because I checked in Odoo and this notation only appears a few times, so we can't expect everyone to understand we want a copy by only reading the code.